File: 450dsa/Graph/BFS.py
from collections import defaultdict

class Graph:
    
    def __init__(self):
        #default dict to store the graph
      self.graph=defaultdict(list)
    
    def addEdge(self,u,v):
        self.graph[u].append(v)
    
    #mark all the vertices are not visited     
    def BFS(self,s):
        visited=[False]*(max(self.graph)+1)
        
    # creating a queue for a graph
        queue=[]
    #mark the source node as visited and enqueue it
        queue.append(s)
        visited[s]=True


        while queue:
          #dequeue a vertex  from queue and simply print it 
            s=queue.pop(0)
            # cout<<s<<" ";
            print(s,end=" ")
           
          
            for i in self.graph[s]: # Get all adjacent vertices of the
                if visited[i]==False:  # dequeued vertex s  If a adjacent has not been visited
                    queue.append(i)   # enqueue
                    visited[i]=True #mark it visited
            
g = Graph()#create object
g.addEdge(0, 1)#pushing 
g.addEdge(0, 2)
g.addEdge(1, 2)
g.addEdge(2, 0)
g.addEdge(2, 3)
g.addEdge(3, 3)
 
print ("Following is Breadth First Traversal"
                  " (starting from vertex 2)")
g.BFS(2)#call the function to perform BFS

File: 450dsa/Graph/BFS.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution
{
    public:
    //Function to return Breadth First Traversal of given graph.
	vector<int>bfsOfGraph(int V, vector<int> adj[])
	{
	    // Code here
	    vector<int>v;
	    queue<int>qu;
	    //pushing source node
	    qu.push(0);
	    
	    vector<int> vis(V,0);
	    vis[0]=1;
	    
	    while(!qu.empty())
	    {
	        int tq=qu.front();
	        qu.pop();
	        v.push_back(tq);
	        
	        for(auto i:adj[tq])
	        {
	            if(!vis[i])
	            {
	                
	                vis[i]=1;
	                qu.push(i);
	            }
	        }
	    }
	    return v;
	}
};

int main()
{
    int V,edge;
    cin>>V>>edge;
    vector<int>adj[V];
    for(int i=0;i<edge;i++)
    {
        int x,y;
        cin>>x>>y;
        adj[x].push_back(y);
        // adj[y].push_back(x);
    }
    
    Solution obj;
    vector<int>res;
    res=obj.bfsOfGraph(V,adj);
    for(int i=0;i<res.size();i++)
    {
        cout<<res[i]<<" ";
    }
    cout<<endl;
    return 0;
}
